File: part1/src/main/java/com/peterlaurence/book/javatokotlin/part1/core/java/PurchasesProviderImpl.java
package com.peterlaurence.book.javatokotlin.part1.core.java;

import com.peterlaurence.book.javatokotlin.part1.fragments.java.PurchasesViewModel.PurchasesProvider;

import java.util.Arrays;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class PurchasesProviderImpl implements PurchasesProvider {
    private ExecutorService executor = Executors.newSingleThreadExecutor();

    @Override
    public void fetchPurchases(String user, PurchaseFetchCallback callback) {
        /* perform asynchronous work */
        executor.submit(() -> {
            try {
                Thread.sleep(1000);
                callback.onPurchaseFetchDone(Arrays.asList("Purchase1", "Purchase2"));
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });
    }
}

File: part1/src/main/java/com/peterlaurence/book/javatokotlin/callbackpattern/core/PurchasesProviderImpl.kt
package com.peterlaurence.book.javatokotlin.callbackpattern.core

import com.peterlaurence.book.javatokotlin.callbackpattern.fragments.PurchasesViewModel.PurchasesProvider
import com.peterlaurence.book.javatokotlin.callbackpattern.fragments.PurchasesViewModel.PurchasesProvider.PurchaseFetchCallback
import java.util.concurrent.Executors

class PurchasesProviderImpl : PurchasesProvider {
    private val executor =
        Executors.newSingleThreadExecutor()

    override fun fetchPurchases(
        user: String,
        callback: PurchaseFetchCallback
    ) {
        /* perform asynchronous work */
        executor.submit {
            try {
                Thread.sleep(1000)
                callback.onPurchaseFetchDone(
                    listOf("Purchase1", "Purchase2")
                )
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
    }
}